File: 0494-target-sum/0494-target-sum.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        unordered_map<string, int> memo;
        return dfs(0, 0, nums, target, memo);

       
    }
    int dfs(int index, int curr, vector<int>& nums, int target, unordered_map<string,int>& memo) {
        if (index == nums.size()) {
            return curr == target ? 1 : 0;
        }

        string key = to_string(index) + "," + to_string(curr);

        if (memo.count(key))
            return memo[key];

        int add = dfs(index + 1, curr + nums[index], nums, target, memo);
        int sub = dfs(index + 1, curr - nums[index], nums, target, memo);

        memo[key] = add + sub;
        return memo[key];
    }
     
};

File: 0494-target-sum/0494-target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:

        memo = {}
        def dfs(index = 0, targetSum = target):
            if(index == len(nums)):
                if(targetSum == 0):
                    return 1
                return 0
            if((index, targetSum) in memo):
                return memo[(index, targetSum)]
            else:
                memo[(index, targetSum)] = dfs(index+1, targetSum + nums[index])+dfs(index+1, targetSum - nums[index])
                return memo[(index, targetSum)]
        return dfs()
        